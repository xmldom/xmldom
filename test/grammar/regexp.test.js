'use strict';

const fs = require('fs');
const { describe, expect, test } = require('@jest/globals');
const grammar = require('../../lib/grammar');
const alphabetical = Object.keys(grammar)
	.filter((key) => grammar[key] instanceof RegExp)
	.sort();
const Grammar = Object.keys(grammar)
	.filter((key) => grammar[key] instanceof RegExp)
	// then by the length (complexity) of the regular expression
	// shortest ones first
	.sort((a, b) => {
		const length = grammar[a].source.length - grammar[b].source.length;
		return length === 0 ? alphabetical.indexOf(a) - alphabetical.indexOf(b) : length;
	})
	.reduce((acc, key) => {
		acc[key] = grammar[key];
		return acc;
	}, {});
var REGEXP_DUMP = `'use strict';
// THIS FILE IS GENERATED by tests, don't change it manually
${Object.entries(Grammar)
	.map(
		([name, reg]) =>
			`const ${name} = /${reg.source
				// to make the test pass with all node version,
				// we need to "sync" how `/` is being serialized
				// in node v10 it serializes to just /
				// in later versions it serializes to \/
				.replace(/\\?\//g, '\\/')}/${reg.flags};`
	)
	.join('\n')}`;
describe('all grammar regular expressions', () => {
	test('should have the expected keys', () => {
		expect(Object.keys(Grammar)).toMatchSnapshot();
	});
	test('should match the file on disk', () => {
		var fileName = __dirname + '/regexp.js';
		// delete the file and rerun the test(s) to update to current value, in case you touched grammar.js
		if (!fs.existsSync(fileName)) {
			fs.writeFileSync(fileName, REGEXP_DUMP);
		}
		expect(fs.readFileSync(fileName, 'utf-8')).toBe(REGEXP_DUMP);
	});
});
