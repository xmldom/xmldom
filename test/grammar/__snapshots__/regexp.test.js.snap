// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all grammar regular expressions should have the expected keys 1`] = `
Array [
  "S",
  "S_OPT",
  "SystemLiteral",
  "Char",
  "CDSect",
  "PubidLiteral",
  "Comment",
  "ExternalID",
  "ExternalID_match",
  "Name",
  "PEReference",
  "XMLDecl",
  "Reference",
  "PI",
  "QName",
  "QName_exact",
  "QName_group",
  "NotationDecl",
  "EntityValue",
  "elementdecl",
  "AttlistDecl",
  "EntityDecl",
]
`;
